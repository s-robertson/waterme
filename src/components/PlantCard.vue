<template>
  <div class="c-plant-card" data-test="plant-card">
    <div class="c-plant-card__options">
      <input
        :id="checkboxId"
        :checked="isChecked"
        type="checkbox"
        @click="handleChange"
        data-test="checkbox-water-plant"
      />
      <label class="c-plant-card__label" :for="checkboxId">
        {{ plant.name }}
      </label>
      <div class="c-plant-card__buttons">
        <router-link :to="`edit-plant/${plant.id}`" class="button--icon">
          <SettingsIcon
            class="c-plant-card__settings"
            :alt="`Edit ${plant.name}`"
          />
        </router-link>
        <button
          class="button--icon plant-card__delete"
          @click.prevent="handleDeleteClick"
          data-test="button-delete-plant"
        >
          <TrashIcon />
          <span class="sr-only">{{ `Delete ${plant.name}` }}</span>
        </button>
      </div>
    </div>
    <div class="c-plant-card__watering-progress">
      <div class="c-plant-card__watering-progress-text">
        {{ `${daysRemaining} Days Remaining` }}
      </div>
      <div
        class="c-plant-card__watering-progress-fill"
        :style="progressStyle"
      />
    </div>
  </div>
</template>

<script>
import { mapActions } from "vuex";
import moment from "moment";
import SettingsIcon from "@/assets/settings.svg";
import TrashIcon from "@/assets/trash.svg";

export default {
  name: "PlantCard",
  props: {
    plant: Object,
    isChecked: Boolean
  },
  computed: {
    checkboxId() {
      return `water-plant-${this.plant.id}`;
    },
    daysRemaining() {
      const daysBetweenWatering = this.plant.days;
      const lastWatered = moment
        .unix(this.plant.lastWatered)
        .set({ second: 0, minute: 0, hour: 0 });
      const now = moment().set({ second: 0, minute: 0, hour: 0 });
      const daysSinceLastWatered = now.diff(lastWatered, "days");
      const daysRemaining = daysBetweenWatering - daysSinceLastWatered;

      if (daysRemaining > daysBetweenWatering) {
        return daysBetweenWatering;
      }

      return daysRemaining > 0 ? daysRemaining : 0;
    },
    wateringProgress() {
      return (this.daysRemaining / this.plant.days) * 100;
    },
    progressStyle() {
      return {
        width: `${this.wateringProgress}%`
      };
    }
  },
  methods: {
    ...mapActions(["deletePlant"]),
    handleDeleteClick() {
      this.deletePlant(this.plant.id);
    },
    handleChange(e) {
      this.$emit("toggle", e.target.checked);
    }
  },
  components: {
    SettingsIcon,
    TrashIcon
  }
};
</script>

<style lang="scss">
@import "../assets/scss/variables";

.c-plant-card__options {
  margin-bottom: 1rem;
}

.c-plant-card__buttons {
  float: right;
  display: flex;
}

.c-plant-card__settings {
  width: 25px;
  cursor: pointer;
  fill: $link;
  margin-right: 0.5rem;
}

.c-plant-card__watering-progress {
  border: 2px solid #fff;
  position: relative;
  height: 2rem;
  border-radius: 10px;
}

.c-plant-card__watering-progress-fill {
  background: $brand-secondary;
  height: 2rem;
  border-radius: 10px;
}

.plant-card__delete {
  width: 25px;
  vertical-align: middle;
  display: inline-block;
}

.c-plant-card__watering-progress-text {
  position: absolute;
  width: 100%;
  text-align: center;
  height: 1rem;
  padding: 0.5rem 0;
  box-sizing: content-box;
}
</style>
