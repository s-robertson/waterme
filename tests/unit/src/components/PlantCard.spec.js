import MockDate from "mockdate";
import moment from "moment";
import { shallowMount } from "@vue/test-utils";
import PlantCard from "@/components/PlantCard";

describe("PlantCard component", () => {
  beforeAll(() => {
    MockDate.set("2020-07-30");
  });

  afterAll(() => {
    MockDate.reset();
  });

  it("Should render correctly", () => {
    const mockPlant = {
      name: "My Plant",
      lastWatered: moment.utc("2020-07-25").unix(),
      days: 10
    };
    const wrapper = shallowMount(PlantCard, {
      propsData: {
        plant: mockPlant
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  const dataset = [
    [7, "2020-07-23", 0],
    [7, "2020-06-20", 0],
    [10, "2020-07-25", 5],
    [10, "2020-08-15", 10],
    [8, "2020-07-30", 8],
    [8, "2020-07-29", 7],
    [1, "2020-07-29", 0],
    [1, "2020-07-30", 1]
  ];

  it.each(dataset)(
    `Should determine correct days remaining for %i days and %s as date`,
    (daysBetweenWatering, mockDate, expectedDaysRemaining) => {
      const mockPlant = {
        name: "My Plant",
        lastWatered: moment.utc(mockDate).unix(),
        days: daysBetweenWatering
      };

      const wrapper = shallowMount(PlantCard, {
        propsData: {
          plant: mockPlant
        }
      });

      expect(wrapper.vm.daysRemaining).toBe(expectedDaysRemaining);
    }
  );
});
