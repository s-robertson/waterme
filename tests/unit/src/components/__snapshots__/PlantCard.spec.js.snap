// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantCard component Should render correctly 1`] = `
<div data-test="plant-card" class="c-plant-card">
  <div class="c-plant-card__options"><input id="water-plant-123" type="checkbox" data-test="checkbox-water-plant"> <label for="water-plant-123" class="c-plant-card__label">
      My Plant
    </label>
    <div class="c-plant-card__buttons">
      <router-link-stub to="edit-plant/123" class="button--icon">
        <settingsicon-stub alt="Edit My Plant" class="c-plant-card__settings"></settingsicon-stub>
      </router-link-stub> <button data-test="button-delete-plant" class="button--icon plant-card__delete">
        <trashicon-stub></trashicon-stub> <span class="sr-only">Delete My Plant</span>
      </button>
    </div>
  </div>
  <div class="c-plant-card__watering-progress">
    <div class="c-plant-card__watering-progress-text">
      5 Days Remaining
    </div>
    <div class="c-plant-card__watering-progress-fill" style="width: 50%;"></div>
  </div>
</div>
`;
